.ci/magician/cmd/scheduled_pr_reminders.go

```diff
@@ -355,15 +346,59 @@ func notificationState(pr *github.PullRequest, issueEvents []*github.IssueEvent,
 	}
 
 	if earliestChangesRequested != nil {
-		return waitingForContributor, *earliestChangesRequested.SubmittedAt.GetTime(), nil
+		timeState := maxTime(*earliestChangesRequested.SubmittedAt.GetTime(), readyForReviewTime)
+		return waitingForContributor, timeState, nil
 	}
 	if earliestApproved != nil {
-		return waitingForMerge, *earliestApproved.SubmittedAt.GetTime(), nil
+		timeState := maxTime(*earliestApproved.SubmittedAt.GetTime(), readyForReviewTime)
+		return waitingForMerge, timeState, nil
 	}
 	if earliestCommented != nil {
-		return waitingForContributor, *earliestCommented.SubmittedAt.GetTime(), nil
+		timeState := maxTime(*earliestCommented.SubmittedAt.GetTime(), readyForReviewTime)
+		return waitingForContributor, timeState, nil
+	}
```

Comment:
We only consider the most recent review from a core reviewer. See the case:switch on line 340.

Reply:
The behavior I described does _theoretically_ exist - here's a test case that demonstrates it:

```go


		"changes_requested before ready_for_review with approved after": {
			pullRequest: &github.PullRequest{
				User:      &github.User{Login: github.String("author")},
				CreatedAt: &github.Timestamp{time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC)},
			},
			issueEvents: []*github.IssueEvent{
				&github.IssueEvent{
					Event:             github.String("review_requested"),
					CreatedAt:         &github.Timestamp{time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC)},
					RequestedReviewer: &github.User{Login: github.String(firstCoreReviewer)},
				},
				&github.IssueEvent{
					Event:     github.String("ready_for_review"),
					CreatedAt: &github.Timestamp{time.Date(2024, 1, 3, 0, 0, 0, 0, time.UTC)}, // Earlier ready
				},
			},
			reviews: []*github.PullRequestReview{
				&github.PullRequestReview{
					User:        &github.User{Login: github.String(firstCoreReviewer)},
					State:       github.String("CHANGES_REQUESTED"),
					SubmittedAt: &github.Timestamp{time.Date(2024, 1, 2, 0, 0, 0, 0, time.UTC)},
				},
				&github.PullRequestReview{
					User:        &github.User{Login: github.String(secondCoreReviewer)},
					State:       github.String("APPROVED"),
					SubmittedAt: &github.Timestamp{time.Date(2024, 1, 4, 0, 0, 0, 0, time.UTC)},
				},
			},
			expectState: waitingForMerge,
			expectSince: time.Date(2024, 1, 4, 0, 0, 0, 0, time.UTC),
		},
```

However, after thinking it over, I think the behavior you have here makes sense... if a change was requested, and then the PR was closed, left for a long time, and then reopened (somehow without a review being requested) we'd want to keep it in "Changes requested" but just reset the clock. (But actually ready_for_review requests a new review, so the only case that can _actually_ happen here is that a review was requested and we reset the clock for when the review should have started.)

Reply:
hmm I think there actually is an issue here with dismissed reviews not clearing an earlier changes_requested.

Theoretically the new reviewer should dismiss the old reviewer if their feedback is no-longer applicable and/or they can no longer approve.

Reply:
Is there an issue with dismissed reviews that was not already present? I don't remember exactly why I implemented this as is; I'm happy to change the behavior around dismissed reviews but I'd rather hold it for a separate PR unless there's a specific impact to dismissed reviews from this work, since it may have additional implications.